Specs/AES.lean

```diff
@@ -0,0 +1,227 @@
+def RotWord (w : BitVec WordSize) : BitVec WordSize :=
+  match_bv w with
+  | [a3:8, a2:8, a1:8, a0:8] => a0 ++ a3 ++ a2 ++ a1
+  | _ => w -- unreachable case
+
+def SubWord (w : BitVec WordSize) : BitVec WordSize :=
+  match_bv w with
+  | [a3:8, a2:8, a1:8, a0:8] => (sbox a3) ++ (sbox a2) ++ (sbox a1) ++ (sbox a0)
+  | _ => w -- unreachable case
+
+protected def InitKey {Param : KBR} (i : Nat) (key : BitVec Param.key_len)
+  (acc : KeySchedule) : KeySchedule :=
+  if h₀ : Param.Nk ≤ i then acc
+  else
+    have h₁ : i * 32 + 32 - 1 - i * 32 + 1 = WordSize := by
+      simp only [WordSize]; omega
+    let wd := h₁ ▸ extractLsb (i * 32 + 32 - 1) (i * 32) key
+    let (x:KeySchedule) := [wd]
+    have _ : Param.Nk - (i + 1) < Param.Nk - i := by omega
+    AES.InitKey (Param := Param) (i + 1) key (acc ++ x)
+  termination_by (Param.Nk - i)
+
+protected def KeyExpansion_helper {Param : KBR} (i : Nat) (ks : KeySchedule)
+  : KeySchedule :=
+  if h : 4 * Param.Nr + 4 ≤ i then
+    ks
+  else
+    let tmp := List.get! ks (i - 1)
+    let tmp :=
+      if i % Param.Nk == 0 then
+        (SubWord (RotWord tmp)) ^^^ (List.get! Rcon $ (i / Param.Nk) - 1)
+      else if Param.Nk > 6 && i % Param.Nk == 4 then
+        SubWord tmp
+      else
+        tmp
+    let res := (List.get! ks (i - Param.Nk)) ^^^ tmp
+    let ks := List.append ks [ res ]
+    have _ : 4 * Param.Nr + 4 - (i + 1) < 4 * Param.Nr + 4 - i := by omega
+    AES.KeyExpansion_helper (Param := Param) (i + 1) ks
+  termination_by (4 * Param.Nr + 4 - i)
+
+def KeyExpansion {Param : KBR} (key : BitVec Param.key_len)
+  : KeySchedule :=
+  let seeded := AES.InitKey (Param := Param) 0 key []
+  AES.KeyExpansion_helper (Param := Param) Param.Nk seeded
+
+def SubBytes {Param : KBR} (state : BitVec Param.block_size)
+  : BitVec Param.block_size :=
+  have h : Param.block_size = 128 := by simp only [Param.h, BlockSize]
+  h ▸ AESCommon.SubBytes (h ▸ state)
+
+def ShiftRows {Param : KBR} (state : BitVec Param.block_size)
+  : BitVec Param.block_size :=
+  have h : Param.block_size = 128 := by simp only [Param.h, BlockSize]
+  h ▸ AESCommon.ShiftRows (h ▸ state)
+
+def XTimes (bv : BitVec 8) : BitVec 8 :=
+  let res := truncate 7 bv ++ 0b0#1
+  if getLsb bv 7 then res ^^^ 0b00011011#8 else res
+
+def MixColumns {Param : KBR} (state : BitVec Param.block_size)
+  : BitVec Param.block_size :=
+  have h : Param.block_size = 128 := by simp only [Param.h, BlockSize]
+  let FFmul02 := fun (x : BitVec 8) => XTimes x
+  let FFmul03 := fun (x : BitVec 8) => x ^^^ XTimes x
+  h ▸ AESCommon.MixColumns (h ▸ state) FFmul02 FFmul03
+
+-- TODO: looks like a SAT/SMT problem
+protected theorem FFmul02_equiv : (fun x => XTimes x) = DPSFP.FFmul02 := by
+  funext x
+  simp only [XTimes, DPSFP.FFmul02]
+  sorry
+
+-- TODO: looks like a SAT/SMT problem
+protected theorem FFmul03_equiv : (fun x => x ^^^ XTimes x) = DPSFP.FFmul03 := by
+  funext x
+  simp only [XTimes, DPSFP.FFmul03]
+  sorry
```

Comment:
I remember that having `sorry` on small lemmas was fine, but to double check: @shigoel  :)

Reply:
Thanks, @aqjune-aws!

Yes indeed --- `sorry` for this particular type of lemma is acceptable for now. This lemma is a great candidate for proof by SAT/SMT solvers, and we can tackle them using those automated decision procedures once those tactics are available.

BTW, LNSym has many such `sorry`s already (e.g., in this [file](https://github.com/leanprover/LNSym/blob/main/Arm/SeparateProofs.lean.)).